### ISSUE TITLE
Incorrect [SC2034: var1 appears unused. Verify it or export it.] 

### ISSUE THREAD

**Issue by Lathoudero** [koalaman/shellcheck#801]
#### For bugs
- SC2034
- online
- [X ] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [X ] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [X ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash 
myFunction()
{
  eval string1="$1"
  eval string2="$2"
  eval string3="$3"

  echo "string1 = ${string1}"
  echo "string2 = ${string2}"
  echo "string3 = ${string3}"
}

var1="firstString"
var2="second string with spaces"
var3="thirdString"

myFunction "\${var1}" "\${var2}" "\${var3}"

exit 0

```

#### Here's what shellcheck currently says:
Line 13:
var1="firstString"
^-- SC2034: var1 appears unused. Verify it or export it.
 
Line 14:
var2="second string with spaces"
^-- SC2034: var2 appears unused. Verify it or export it.
 
Line 15:
var3="thirdString"
^-- SC2034: var3 appears unused. Verify it or export it.

#### Here's what I wanted or expected to see:

$ shellcheck myscript
No issues detected!

Because vars are used:
myFunction "\${var1}" "\${var2}" "\${var3}"

**Comment by brother** [comment]
It's a bit strange that bash actually uses those vars when you try to escape the dollar sign in the string. My first guess would be that shellcheck was correct but it's indeed not.

If you change the double quotes to single quotes it get even worse.

And that usage of curly braces and eval and what not is also a bit annoying as it makes the "local" variables string1 and so on not assigned according to shellcheck 0.4.5 (some days ago...need to rebuild).

Interesting find all in all.

**Comment by koalaman** [comment]
Unfortunately there's no general way to statically determine whether a variable will ever be assigned or accessed through indirections like `eval` for non-trivial cases. 

A statement like `eval string1="$1"` may or may not result in assignments to `string1` or `arbitraryOtherVar`, and it may or may not access some variable `var1`, depending on the runtime value of `$1`. ShellCheck doesn't try to guess.
